CRM/Core/BAO/FinancialTrxn.php

```diff
@@ -1,742 +1,8 @@
-   *
-   */
-  public static function updateFinancialAccountsOnPaymentInstrumentChange($inputParams) {
-    $prevContribution = $inputParams['prevContribution'];
-    $currentContribution = $inputParams['contribution'];
-    // ensure that there are all the information in updated contribution object identified by $currentContribution
-    $currentContribution->find(TRUE);
-
-    $deferredFinancialAccount = $inputParams['deferred_financial_account_id'] ?? NULL;
-    if (empty($deferredFinancialAccount)) {
-      $deferredFinancialAccount = CRM_Contribute_PseudoConstant::getRelationalFinancialAccount($prevContribution->financial_type_id, 'Deferred Revenue Account is');
-    }
-
-    $lastFinancialTrxnId = self::getFinancialTrxnId($prevContribution->id, 'DESC', FALSE, NULL, $deferredFinancialAccount);
-
-    // there is no point to proceed as we can't find the last payment made
-    // @todo we should throw an exception here rather than return false.
-    if (empty($lastFinancialTrxnId['financialTrxnId'])) {
-      return FALSE;
-    }
-
-    // If payment instrument is changed reverse the last payment
-    //  in terms of reversing financial item and trxn
-    $lastFinancialTrxn = civicrm_api3('FinancialTrxn', 'getsingle', ['id' => $lastFinancialTrxnId['financialTrxnId']]);
-    unset($lastFinancialTrxn['id']);
-    $lastFinancialTrxn['trxn_date'] = $inputParams['trxnParams']['trxn_date'];
-    $lastFinancialTrxn['total_amount'] = -$inputParams['trxnParams']['total_amount'];
-    $lastFinancialTrxn['net_amount'] = -$inputParams['trxnParams']['net_amount'];
-    $lastFinancialTrxn['fee_amount'] = -$inputParams['trxnParams']['fee_amount'];
-    $lastFinancialTrxn['contribution_id'] = $prevContribution->id;
-    foreach ([$lastFinancialTrxn, $inputParams['trxnParams']] as $financialTrxnParams) {
-      $trxn = CRM_Core_BAO_FinancialTrxn::create($financialTrxnParams);
-      $trxnParams = [
-        'total_amount' => $trxn->total_amount,
-        'contribution_id' => $currentContribution->id,
-      ];
-      CRM_Contribute_BAO_Contribution::assignProportionalLineItems($trxnParams, $trxn->id, $prevContribution->total_amount);
-    }
-
-    self::createDeferredTrxn(CRM_Utils_Array::value('line_item', $inputParams), $currentContribution, TRUE, 'changePaymentInstrument');
-
-    return TRUE;
-  }
-
-  /**
-   * Generate and assign an arbitrary value to a field of a test object.
-   *
-   * Always set is_payment to 1 as this is used for Payment api as  well as FinancialTrxn.
-   *
-   * @param string $fieldName
-   * @param array $fieldDef
-   * @param int $counter
-   *   The globally-unique ID of the test object.
-   */
-  protected function assignTestValue($fieldName, &$fieldDef, $counter) {
-    if ($fieldName === 'is_payment') {
-      $this->is_payment = 1;
-    }
-    else {
-      parent::assignTestValue($fieldName, $fieldDef, $counter);
-    }
-  }
-
-}
+class CRM_Core_BAO_FinancialTrxn extends CRM_Financial_BAO_FinancialTrxn {}
```

Comment:
We're in kind of uncharted territory with renaming a DAO/BAO (eg there are various bits which map back/forth between the naming types).

I would hazard a guess that class-aliasing would work better than subclassing:

```php
// Subclassing
class CRM_Core_DAO_FinancialTrxn extends CRM_Financial_DAO_FinancialTrxn {}
class CRM_Core_BAO_FinancialTrxn extends CRM_Financial_BAO_FinancialTrxn {}

// Class aliasing
class_alias('CRM_Financial_DAO_FinancialTrxn', 'CRM_Core_DAO_FinancialTrxn');
class_alias('CRM_Financial_BAO_FinancialTrxn', 'CRM_Core_BAO_FinancialTrxn');
```

I haven' tested this at all... but one would expect these to differ in handling static data - eg `static::$foo` or `Civi::$statics[__CLASS__]['foo']`:
* In subclassing, maybe more static structures split apart (`Civi::$statics['CRM_OLD']` vs `Civi::$statics['CRM_NEW']`).
* In aliasing, maybe more of the static structures stay together (`Civi::$statics['CRM_NEW']`).

Reply:
Hmm - it's not really unchartered - it would be about number 8ish I think. I don't think that is the test fail cause though

Reply:
OK - I have the cause of the test fails - because the BAO name is currently wrong the apiv4 is actually not correctly figuring out the create method name & is calling writeObjects not create. The upshot is that we are doing something bad now. When we fix it the test fails because the test is not configured with all the required sample data

Reply:
One fix would be for the API to override this method:
https://github.com/civicrm/civicrm-core/blob/5a443458b6e14de9c21ded8cc3f1976228a47524/Civi/Api4/Generic/Traits/DAOActionTrait.php#L38-L43

Reply:
@colemanw @pradpnayak's PR does an OK job of the rename & he has incorportated your cleanup into #22400 - however, it turns out the BAO naming issue means the v4 api has been bypassing the `create` function - which would be really bad if we thought people were using it much in real life.

Part of the fix is to add the spec function to v4 which I've done in #22571 - but tests still need fixing to cope with the new spec requirements

This PR doesn't add anything over & above the 2 mentioned above so closing it